src/coreclr/src/jit/lower.cpp

```diff
@@ -3024,6 +3056,132 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
+                    GenTreeLclVar* lclVar = retVal->AsLclVar();
+                    LclVarDsc*     varDsc = comp->lvaGetDesc(lclVar);
+                    assert(varDsc->lvIsStructField);
+                }
+                else
+                {
+                    assert(retVal->OperIs(GT_LCL_FLD));
+                    // 'special' helpers that return true from `TreatAsHasRetBufArg`
+                    // and can create trees like:
+                    // N004 (  3,  2) [000018] -------N----        t18 =    LCL_VAR_ADDR byref  V05 tmp1
+                    //                                                   *    bool   V05.hasValue (offs=0x00) -> V08
+                    //                                                   tmp4
+                    //                                                   *    bool   V05.value (offs=0x01) -> V09 tmp5
+                    // N006 (  3,  2) [000015] ------------        t15 =    LCL_VAR   ref   (AX) V02 loc0          $380
+                    // N007 (  3, 10) [000016] ------------        t16 =    CNS_INT(h) long   0xd1ffab1e class $1c1
+                    //                                                   /--*  t18    byref  arg0 in rcx
+                    //                                                   +--*  t15    ref    arg2 in r8
+                    //                                                   +--*  t16    long   arg1 in rdx
+                    // N008 ( 23, 23) [000017] --CXG-------              *  CALL help void
+                    // HELPER.CORINFO_HELP_UNBOX_NULLABLE $206
+                    // N001 (  4,  5) [000020] ------------        t20 =    LCL_FLD   short  V05 tmp1         [+0]
+                    //                                                   *    bool   V05.hasValue (offs=0x00) -> V08
+                    //                                                   tmp4
+                    //                                                   *    bool   V05.value (offs=0x01) -> V09 tmp5
+                    //                                                   $400
+                    //                                                   /--*  t20    short
+                    // N002 (  5,  6) [000021] ----G-------              *  RETURN    struct $2c2
+                    // TODO seandree: find a way to get rid of them.
+                }
+                assert(genActualType(comp->info.compRetNativeType) == genActualType(retVal->TypeGet()));
+            }
+        }
+#endif
+        if (varTypeIsStruct(ret))
+        {
+            assert(comp->compNoReturnRetyping());
+            GenTree*  retVal           = ret->gtGetOp1();
+            var_types nativeReturnType = genActualType(comp->info.compRetNativeType);
+
+            ret->ChangeType(nativeReturnType);
+
+            assert(retVal->OperIs(GT_IND, GT_OBJ, GT_LCL_VAR, GT_LCL_FLD, GT_BITCAST, GT_CNS_INT, GT_SIMD));
+            switch (retVal->OperGet())
```

Comment:
For the non-indirection cases, I think you should be able to handle them more uniformly - that is, if the register file type of the source is the same as the return type, then just retype the source. Otherwise add a `BITCAST`.